### PR TITLE
Wait to start loading source assets until `onAdd` is called

### PR DESCRIPTION
**Pull request by lbud**
Fixes https://github.com/mapbox/mapbox-gl-js/issues/3735 by taking advantage of the `onAdd` method to load assets/fire events. Previously we were doing those in the source constructors, and they were constructed before establishing ownership (map -> style -> source_cache -> source), meaning events fired almost immediately wouldn't bubble up to the map object — so for example, vector tile sources (loaded over network connection) would reliably trigger `map.on('source.load', fn)` listeners, but canvas sources which loaded very quickly (from the same page) would not.

This isn't a perfect solution, but since sources will soon undergo a pretty big overhaul… 🔨 

cc @lucaswoj

### CODE REVIEW

File: js/style/style.js [hunk 1]
```diff
@@ -374,14 +373,14 @@ class Style extends Evented {
...
             isSourceLoaded: sourceCache.loaded(),
             source: sourceCache.serialize(),
             sourceId: id
         }));
-
+        sourceCache.style = this;
         sourceCache.onAdd(this.map);
+
         this._changed = true;
     }
```
Comment (lucaswoj) [inline js/style/style.js]: Are the changes to this file significant?

**Comment by lbud** [comment]
@lucaswoj the way it has been working:
* `Style` creates instances of `SourceCache`s, _then_ sets their parent: [style.js#L376-L382](https://github.com/mapbox/mapbox-gl-js/blob/2d10449743a61456ed3585ad97a101ad6252f6e8/js/style/style.js#L376-L382)
  * it does this because its `eventedParentData` is dependent on the Source having already been created (`sourceCache.serialize()` and `sourceCache.loaded()` both just forward to `sourceCache._source` methods)
  * this is where the race occurs: when sources fire events in their constructors before we've set up SourceCache's parent. 
    * Specifically I was seeing this with the canvas source branch, since we load so quickly (from the same page). 
    * However, at a glance I can see that other events like [`VectorTileSource#dataloading`](https://github.com/mapbox/mapbox-gl-js/blob/2d10449743a61456ed3585ad97a101ad6252f6e8/js/source/vector_tile_source.js#L31) wouldn't always bubble up to `Map` for the same reason. Was just able to reproduce this with dynamically adding a vector tile set (it fires [`source.load`](https://github.com/mapbox/mapbox-gl-js/blob/2d10449743a61456ed3585ad97a101ad6252f6e8/js/source/vector_tile_source.js#L40) because it's slow enough, but never [`dataloading`](https://github.com/mapbox/mapbox-gl-js/blob/2d10449743a61456ed3585ad97a101ad6252f6e8/js/source/vector_tile_source.js#L31))
* `SourceCache` creates instances of `Source`s, passing itself as a constructor so a source can be responsible for calling `setEventedParent`
  * We were unnecessarily setting this twice: in individual source constructors and also in `Source.create` https://github.com/mapbox/mapbox-gl-js/pull/3761/files#diff-d5506f072d1241401a7e31ff642c1606L24

Another minor, related change I made in this PR is to pass `Style` in the `ImageSprite` constructor, which looked like it could feasibly suffer the same race condition: https://github.com/mapbox/mapbox-gl-js/pull/3761/files#diff-6369a555a491e3e9a58fd29c3fd5a371L97

**Comment by lucaswoj** [comment]
@lbud Are you ready to merge this PR? I'd hate for a rebase conflict to emerge 😅

**Comment by lbud** [comment]
Oops yes @lucaswoj — I paused on this with all my naïve optimism about #3186 😬 
🚢

**Pull request merged**